### PR TITLE
Refactor OAuthDeviceCode to support non-Entra IdPs

### PR DESCRIPTION
**Pull request by gregertw**
## Description

1. Added support for non-Microsoft Entra identity providers:

* New oauth_discovery_url parameter for standard OAuth discovery URLs
* Disabled instance discovery and authority validation for non-Entra authorities

2. Enhanced token caching:

* Added clear_cache option to clear existing token cache
* Introduced mem_cache_only flag for in-memory-only token caching

3. Improved token refresh logic:

* Now checks for valid refresh tokens before initiating a new device code flow

4. Added new convenience methods:

* OAuthDeviceCode.default() for generic identity providers
* OAuthDeviceCode.default_for_azure_ad() for Azure AD with pre-configured settings

5. Updated error handling and user prompts for device code flow

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

**Review by doctrino — COMMENTED** [review]
Tried the flow and fails with the following error
```bash
cognite.client.exceptions.CogniteAuthError: Error generating access token! Error: invalid_client, error description: AADSTS7000218: The request body must contain the following parameter: 'client_assertion' or 'client_secret'. Trace ID: f0fb6991-ed32-4acd-953f-d8f5f333fe00 Correlation ID: 7d3ffed0-ee7d-4412-ba3a-b3a0915947ad Timestamp: 2024-09-02 14:32:11
```

In addition, some nitpick.

File: cognite/client/credentials.py [hunk 1]
```diff
@@ -163,23 +188,50 @@ class OAuthDeviceCode(_OAuthCredentialProviderWithTokenRefresh, _WithMsalSeriali
...
+        scopes: list[str] = ["IDENTITY user_impersonation profile openid"],
+        oauth_discovery_url: str | None = None,
         token_cache_path: Path | None = None,
         token_expiry_leeway_seconds: int = _TOKEN_EXPIRY_LEEWAY_SECONDS_DEFAULT,
+        clear_cache: bool = False,
+        mem_cache_only: bool = False,
+        **token_custom_args: Any,
     ) -> None:
         super().__init__(token_expiry_leeway_seconds)
+        if not authority_url and not oauth_discovery_url:
```
[the remark this answers is not in the document]
Comment (haakonvt) [reply cognite/client/credentials.py:213]: ...there's also `at_least_one_is_not_none` and `at_most_one_is_not_none`

File: cognite/client/credentials.py [hunk 2]
```diff
@@ -242,23 +278,53 @@ class OAuthDeviceCode(_OAuthCredentialProviderWithTokenRefresh, _WithMsalSeriali
...
+        if not at_least_one_is_not_none(scopes, cdf_cluster):
+            raise ValueError("Either 'scopes' or 'cdf_cluster' must be provided.")
+        if not client_id:
+            raise ValueError("'client_id' must be provided.")
         self.__authority_url = authority_url
+        self.__oauth_discovery_url = oauth_discovery_url
         self.__client_id = client_id
-        self.__scopes = scopes
+        self.__scopes = scopes or [
+            f"https://{cdf_cluster}.cognitedata.com/IDENTITY https://{cdf_cluster}.cognitedata.com/user_impersonation openid profile"
```
Comment (doctrino) [inline cognite/client/credentials.py:314]: @gregertw The original default did not work for me

File: cognite/client/credentials.py [hunk 3]
```diff
@@ -283,22 +360,100 @@ def client_id(self) -> str:
     def scopes(self) -> list[str]:
         return self.__scopes
 
+    def scope_string(self) -> str:
+        return " ".join(self.__scopes)
+
+    def _get_token(self, convert_timestamps: bool = True) -> dict[str, Any]:
```
Comment (doctrino) [inline cognite/client/credentials.py:366]: Added this as a private method to make it possible to obtain the token

**Review by doctrino — APPROVED** [review]

**Pull request merged**